### PR TITLE
fix(ingest/bigquery): Fix for table cache was not cleared 

### PR DESCRIPTION
**Pull request by treff7es**
- Fix for not clearing BigQuery table cache, which caused to carry over tables from one project to the other
- Fixing range partition profiling
- Fixing peak memory report

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/run/pipeline.py [hunk 1]
```diff
@@ -123,15 +123,15 @@ class CliReport(Report):
     py_version: str = sys.version
     py_exec_path: str = sys.executable
     os_details: str = platform.platform()
-    _peek_memory_usage: int = 0
+    _peak_memory_usage: int = 0
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/run/pipeline.py:126]: Nice :)

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py [hunk 2]
```diff
@@ -572,7 +568,10 @@ def _process_project(
                 self.report.report_dropped(f"{bigquery_dataset.name}.*")
                 continue
             try:
-                yield from self._process_schema(conn, project_id, bigquery_dataset)
+                # db_tables and db_views are populated in the this method
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py:571]: Thanks for adding this comment

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py [hunk 3]
```diff
@@ -631,20 +632,27 @@ def _process_project(
...
+    def generate_lineage(
+        self,
+        project_id: str,
+        db_tables: Dict[str, List[BigqueryTable]],
+        db_views: Dict[str, List[BigqueryView]],
+    ) -> Iterable[MetadataWorkUnit]:
         logger.info(f"Generate lineage for {project_id}")
-        for dataset in self.db_tables:
-            for table in self.db_tables[dataset]:
+        for dataset in db_tables:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py:654]: Why not use .keys() here too? Like we do below. Whichever way we go let's keep consistency

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py [hunk 4]
```diff
@@ -768,12 +789,12 @@ def _process_table(
             )
 
         # If table has time partitioning, set the data type of the partitioning field
-        if table.time_partitioning:
-            table.time_partitioning.column = next(
+        if table.partition_info:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py:792]: Curious -- Why was the previous code wrong??

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py [hunk 5]
```diff
@@ -1096,106 +1117,74 @@ def get_tables_for_dataset(
...
+
+                # For sharded tables we only process the latest shard
+                # which has the highest date in the table name.
+                # Sharded tables look like: table_20220120
+                # We only has one special case where the table name is a date
+                # in this case we merge all these tables under dataset name as table name.
+                # For example some_dataset.20220110 will be turned to some_dataset.some_dataset
+                # It seems like there are some bigquery user who uses this way the tables.
+                if shard:
+                    if not sharded_tables.get(table_identifier.get_table_name()):
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py:1154]: This seems dangerous. Why not use the existing table_name variable defined just above? We go on to update the map with the above variable as the key

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py [hunk 6]
```diff
@@ -1096,106 +1117,74 @@ def get_tables_for_dataset(
...
-                # we won't hit too many tables queried with partitions system view.
-                # The key in the map is the actual underlying table name and not the friendly name and
-                # that's why we need to get the actual table names and not the normalized ones.
-                table_items.update(
-                    {value.table_id: value for value in sharded_tables.values()}
-                )
-
-                if table_items:
+                        # When table is none, we use dataset_name as table_name
+                        table_name = table_identifier.get_table_name().split(".")[-1]
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py:1171]: This is already on line 44 above?

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py [hunk 7]
```diff
@@ -1096,106 +1117,74 @@ def get_tables_for_dataset(
...
+                        assert stored_shard
+                        if stored_shard < shard:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py:1173]: This condition makes very little sense to me as an unfamiliar reader. Consider adding a comment to explain

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py [hunk 8]
```diff
@@ -1096,106 +1117,74 @@ def get_tables_for_dataset(
...
+                            sharded_tables[table_name] = table
+                        continue
+                else:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py:1176]: Hard to follow all of these nested if else conditions. Would maybe be easier to read if we broke into small well-named functions

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py [hunk 9]
```diff
@@ -1096,106 +1117,74 @@ def get_tables_for_dataset(
...
+                    table_count = table_count + 1
+                    table_items[table.table_id] = table
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py:1178]: I have no clue from reading the name of this data structure what is inside of it. 'Table items' is a super generic name. Wondering if we can make it clearer

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py [hunk 10]
```diff
@@ -1205,31 +1194,46 @@ def get_tables_for_dataset(
...
+            if table_items:
+                bigquery_tables.extend(
+                    BigQueryDataDictionary.get_tables_for_dataset(
+                        conn,
+                        project_id,
+                        dataset_name,
+                        table_items,
+                        with_data_read_permission=self.config.profiling.enabled,
+                    )
+                )
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py:1218]: so having read through this section of the code I can say I have little idea what the implications of making these changes will be

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py [hunk 11]
```diff
@@ -365,10 +396,8 @@ def get_tables_for_dataset(
                 ddl=table.ddl,
                 expires=tables[table.table_name].expires if tables else None,
                 labels=tables[table.table_name].labels if tables else None,
-                time_partitioning=PartitionInfo.from_time_partitioning(
-                    tables[table.table_name].time_partitioning
-                )
-                if tables and tables[table.table_name].time_partitioning
+                partition_info=PartitionInfo.from_table_info(tables[table.table_name])
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:399]: So we previously assumed all Partitioned Tables were time partitioned? And now we know there's another default case of range partitioning that we are handling?

**Review by jjoyce0510 — COMMENTED** [review]
Overall looks okay to me, but I am definitely nervous about the time partitioning changes given that we've had to change it a few times now

File: metadata-ingestion/tests/unit/test_bigquery_profiler.py [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+from datahub.ingestion.source.bigquery_v2.bigquery_report import BigQueryV2Report
+from datahub.ingestion.source.bigquery_v2.bigquery_schema import (
+    BigqueryColumn,
+    BigqueryTable,
+    PartitionInfo,
+)
+from datahub.ingestion.source.bigquery_v2.profiler import BigqueryProfiler
+
+
+def test_not_generate_partition_profiler_query_if_not_partitioned_sharded_table():
```
Comment (jjoyce0510) [inline metadata-ingestion/tests/unit/test_bigquery_profiler.py:13]: Nice test!

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py (hunk 3 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py]: Fixed

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py (hunk 5 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py]: Fixed

File: metadata-ingestion/tests/unit/test_bigquery_source.py [hunk 13]
```diff
@@ -184,9 +188,78 @@ def test_upstream_table_generation_with_temporary_table_with_multiple_temp_upstr
...
+    client_mock.list_tables.return_value = tableListItems
+    data_dictionary_mock.get_tables_for_dataset.return_value = None
+
+    source = BigqueryV2Source(config=config, ctx=PipelineContext(run_id="test"))
+
+    _ = source.get_tables_for_dataset(
+        conn=client_mock, project_id="test-project", dataset_name="test-dataset"
+    )
+
+    assert data_dictionary_mock.call_count == 1
```
Comment (jjoyce0510) [inline metadata-ingestion/tests/unit/test_bigquery_source.py:258]: Thank you for adding these!

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py (hunk 11 above)
Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:399]: got it - makes sense

**Review by jjoyce0510 — APPROVED** [review]
Thanks for addressing comments!
